### PR TITLE
Ensure JTAG clock idles low in bitbanging mode

### PR DESCRIPTION
**Pull request by adamgreig**
In the current firmware, the bitbang mode leaves JTAG TCK high if it returns early (i.e. the transfer is not a multiple of 8 bits). This can lead to glitches on TCK when the next transfer swaps to SPI mode.

One option is to insert a call to `tck.set_low()` before the early return in `transfer_rw()` and `transfer_wo()`, but instead this PR rearranges the TCK calls to always only pulse it high for each bit, to the same effect.

Here we set TDI, then delay a half-period, rise the clock, delay another half-period, sample TDO (when applicable), and fall the clock, then repeat. This means the target has a full half period of setup and hold around the clock rising edge it samples on. We sample TDO immediately before falling the clock, and the target then sets TDO after the clock falling edge.

I've tested this at 3kHz to 30MHz reading SPI flash through an ECP5 over JTAG and it resolves the glitch I was seeing before.

We could move TDO capture to the clock rising edge instead of just before the clock falling edge, but I don't think there's any advantage, since we can guarantee we'll finish sampling TDO before the target updates it.

### CODE REVIEW

**Comment by korken89** [comment]
Looking good, thanks for fixing!
Can you remove the binary you have committed? Then I'll merge :)

File: firmware/src/jtag.rs [hunk 1]
```diff
@@ -244,18 +243,19 @@ impl<'a> JTAG<'a> {
                     return;
                 }
 
-                // Set TDI, read TDO, and toggle TCK.
+                // We set TDI half a period before the clock rising edge where it is sampled
+                // by the target, and we sample TDO immediately before the clock falling edge
+                // where it is updated by the target.
```
Comment (Disasm) [inline firmware/src/jtag.rs:248]: I'm not sure this TDO sampling approach is correct, but I also don't see when it could cause problems.

**Comment by adamgreig** [comment]
> Can you remove the binary you have committed? Then I'll merge :)

Oops, sorry! Missed that. Removed.

File: firmware/src/jtag.rs (hunk 1 above)
Comment (adamgreig) [reply firmware/src/jtag.rs:248]: The JTAG spec just says that "Changes in the state of the signal driven through TDO shall occur only on the falling edge of either TCK or the optional TRST", without any requirement on when it's sampled. This way we give ourselves the maximum possible time for TDO to settle. In a hardware implementation you'd want to sample on the rising edge in synchronous logic, but in software we can just as easily wait longer.

Still I agree it looks a bit unusual compared to sampling on the rising edge! I can't think of any reasons this wouldn't be fine, but we could sample on the rising edge just to be conservative.

Comment (Disasm) [reply firmware/src/jtag.rs:248]: Ah, I see. Thanks for clarifying!

**Review by Disasm — APPROVED** [review]

**Pull request merged**